### ISSUE TITLE
Pending class renames/package moves - discussion

### ISSUE THREAD

**Issue by kashike** [ModCoderPack/MCPBot-Issues#576]
Discussion for #514.

Here's a work-in-progress set of planned mass renames/moves for 1.13: https://gist.github.com/kashike/08cbb4ba5c1ca8b1f46347fa84d961cd.

Summary:
 - [ ] Smurf naming
   - [ ] TileEntitySpecialRenderers are now suffixed with `TileEntityRenderer` instead of prefixed with `TileEntity` and suffixed with `Renderer`
   - [ ] TileEntities are now suffixed with `TileEntity` instead of prefixed with `TileEntity`
   - [ ] Blocks are now suffixed with `Block` instead of prefixed with `Block`
   - [ ] Commands are now suffixed with `Command` instead of prefixed with `Command`
   - [ ] Enchantments are now suffixed with `Enchantment` instead of prefixed with `Enchantment`
   - [ ] Items are now suffixed with `Item` instead of prefixed with `Item`
   - [ ] Materials are now suffixed with `Material` instead of prefixed with `Material`
   - [ ] Models are now suffixed with `Model` instead of prefixed with `Model`
   - [ ] Particles are now suffixed with `Particle` instead of prefixed with `Particle`
   - [ ] Properties are now suffixed with `Property` instead of prefixed with `Property`
   - [ ] The `Enum` prefix has been dropped from all enumerations.

**Comment by LexManos** [comment]
With 1.13's changes to commands and other large swafts of MC code. Taking this time to do some long standing house cleaning would be useful.
To this end i've spoken to kashike and decided on a few basic things.
The main one being 'De-smurfing' of the classes. Finally changing the names that we had in place from the old net.minecraft.src days. Where things being clumped together by prefix was good. In this aspect. They are now collected together in packages. So having them named with the prefix can change. Tho, due to not wanting to cause class name conflicts in things that will likely be be used together. We had the idea of putting the 'Smurf' as the Suffix instead of the Prefix. 
Exa:
BlockCocoa-> CocoaBlock. 
ItenCocoa -> CocoaItem

Same goes for Entities, TileEntites, Items. Maybe GuiScreen?

Things that have been discussed and rejected:
TileEntity->BlockEntity. I know everyone's arguments for and against this. It is of my opinion that TileEntity is such a deeply ingrained name in the Minecraft modding community that changing it would not be worth it.

**Comment by williewillus** [comment]
`DerivedWorldInfo` is still a pretty bad name though because it gives 0 indication what the class actually does (which is provide an unmodifiable delegate to another worldInfo). I propose `DelegatedWorldInfo` or some variant thereof

**Comment by bs2609** [comment]
From a quick look at the gist so far:

* `net/minecraft/block/properties/PropertyEnum` `net/minecraft/block/properties/Property`
 This should stay as `PropertyEnum`.

* `net/minecraft/world/EnumSkyBlock` `net/minecraft/world/SkyBlock`
 Should be `LightType` (#402)

* `BlockDoor$DoorHalf`, `BlockDoublePlant$BlockHalf`, `BlockBed$PartType`, `BlockPistonExtension$PistonType` and possibly other inner classes can be shortened.

**Comment by williewillus** [comment]
also seems like some nested enums didn't lose their prefixes properly

**Comment by InsomniaKitten** [comment]
`net/minecraft/item/ItemAppleGoldEnchanted` -> `net/minecraft/item/AppleGoldEnchantedItem`

This makes little to no sense to me, personally. Maybe some general naming cleanup should be done too, alongside the shifting from prefix to suffix. This case in particular would make more sense as `EnchantedGoldAppleItem`, imo

**Comment by Shadows-of-Fire** [comment]
I don't agree with moving the `Block` and `Item` prefixes at all.  Just about all mods follow that naming convention, and the class names just feel worse to read if they are moved to suffixes.  The `Enum` prefixes however, should be removed.

**Comment by Leviathan143** [comment]
What exactly is cleaner about suffixing instead of prefixing? It seems rather arbitrary. I see no benefits to it, and since there's no working remapper that I know of, I see the main result of this change being a lot of manual renaming work.

**Comment by Pokechu22** [comment]
What about packets?  They also have prefixes, which don't necessarily need to exist anymore (with the packages) but still have useful information (the first thing you need to know when looking at a packet is whether it's from the client/server)

**Comment by LexManos** [comment]
> What exactly is cleaner about suffixing instead of prefixing? It seems rather arbitrary. I see no benefits to it, and since there's no working remapper that I know of, I see the main result of this change being a lot of manual renaming work.
Personal preferences, and a few other things. Which is why I wanted to get community input. Honestly I prefer the thought of being able to use code completion simpler in my IDE, not having to type the prefix means I can so blocks.S{tab} and get SandStoneBlock instead of needing to do block.BlockS{tab}
As for a renamer: https://github.com/MinecraftForge/Remapper

Packets, My personal option would be to keep the S/C prefix, but we can move the 'Packet' part of the prefix.

**Comment by AtomicBlom** [comment]
I Agree that suffix or prefix seems largely arbitrary, but it does need to be there to differentiate between Block/Item/TileEntity.

Personally, I prefer the suffix because it reads more naturally, when I'm talking about a Gold Block or a Gold Ingot Item, so I'm all for the change.

**Comment by thiakil** [comment]
I prefer prefixes. I go 'I need the Block of type blah', IDE class locating (`ctrl-n` in IDEA) works well with prefixes. As for reading naturally, I'm used to other European languages that have the noun before the adjective.

I also see it as an already accepted standard, most projects I look at use them as prefixes.

**Comment by tterrag1098** [comment]
I will also throw my hat into the "Prefixes were fine and useful for quick lookups" ring. This seems like change for change sake, what is the actual benefit to using suffixes instead?

**Comment by Leviathan143** [comment]
> As for a renamer: https://github.com/MinecraftForge/Remapper

I tried that once, I built it from source and it didn't work. I asked for help and got linked to a binary, which worked better than the source compiled version, but still didn't work. I asked for help again and got nothing, so I gave up and used regexes.

**Comment by Corosauce** [comment]
Agreed with tterrag, the changing of Block and Item based classes from prefix to suffix use seems needless.

**Comment by HenryLoenwind** [comment]
I also prefer the prefixes, because I usually look them up by "BlockXY<shift-space>"/"ItemXY<shift-space>". However, it's not that often that I need to look up vanilla block/item classes anymore. And when I need to, then I usually go via Blocks.XY/Items.XY because half of the block/item data is in the construction, not the class itself.

What I do like is the Enum prefix. Here Enum<shift-space> gives a nice list of all available enums without having to remember how exactly one was named.

**Comment by LexManos** [comment]
>I tried that once, I built it from source and it didn't work. I asked for help and got linked to a binary, which worked better than the source compiled version, but still didn't work. I asked for help again and got nothing, so I gave up and used regexes.

Feel free to file issue reports. Its just a wrapper around the Srg2Source project that is the core to how Forge's patches are generated. So it *should* be fine. If not report a issue properly. And we'll look into it.

As for the comment on the Enum prefix. To be honest removing it would be nice, I also want to remove a lot of the access from EnumHelper. As 1) J9 will screw with it 2) Bulk of Enum's will never actually need to be edited.

**Comment by cpw** [comment]
We should kill TileEntity and go with what Mojang has named them since a long time - BlockEntity.

**Comment by maruohon** [comment]
I don't like the idea of renaming the blocks and items, at all. After many years of modding, the convention of having the "type" as the prefix is very deeply engraved into my brain. And IMO it's also easier/quicker to see what is being talked about, when the "type" ie. Block or Item is the first thing you see when you read the class name. Also when searching for things in the IDE, having the common prefix first should mean less of unwanted things showing up, when I type for example `BlockS`.

All in all, this seems like change for the sake of change, it would introduce a discontinuity in naming conventions and thus "break" all tutorials up until now, and also introduce a naming convention discrepancy between vanilla/Forge and all mods that have adopted the prefix naming convention up until now (which I believe is most mods that I've looked at?). Personally I see zero benefit in this, only downsides and added work and annoyance.

**Comment by LexManos** [comment]
>discrepancy between vanilla/Forge 

?

**Comment by maruohon** [comment]
I meant that vanilla and Forge would use a different naming convention than all/most mods made up to this point. Unless everyone renames all their classes...

**Comment by LexManos** [comment]
Other modders naming conventions don't really effect ours. People are free to name/write their own code how they want to. I am weary of having any decisions we make effected by 'modders code' because the large bulk of 'modders code' I've seen have bad decompiled artifacts in their own written code. So ya, I understand your concern, But i'm not sure we should put much weight into it.

**Comment by HenryLoenwind** [comment]
Just a little half-OT anecdote that came to my mind reading mauohon's comment: When I started on Ender IO most classes were postfixed with their Block/Item/Tile descriptor. Now all of them are prefixed. Why? Because in the navigation tree, this is actually useful: 
![image](https://user-images.githubusercontent.com/1485873/33354369-df776d66-d4b2-11e7-8085-e54569736167.png)

Not really applicable for vanilla classes, unless they get sorted into per-block packages, too. But a good reason why mods would want to stay with prefixes for their own classes.

**Comment by KingLemming** [comment]
While that's true, a LOT of modders write their code in a way to mirror the syntax/structure provided, so a change here does matter.

And I agree with Henry's comment - prefix helps for organization.

**Comment by LexManos** [comment]
Thats the thing tho, Henry, that makes perfect sense if you organize your packages that way. But on the other hand vanilla is the exact opposite. The package is the prefix. That's the main drive behind this. Block classes are ALREADY prefixed with 'block'. I'm still not sure if we are going to do this.

**Comment by tterrag1098** [comment]
They are, but that doesn't help in quick-lookup scenarios, where it's just looking at the simple name, unless you type out the entire FQ name every time.

**Comment by HenryLoenwind** [comment]
BTW, it may be different for you, but for me "block."<ctrl-space> doesn't do a thing. If that'd work, then yes, it'd make no difference ("Block"<ctrl-space> vs "block."<ctrl-space>).

BTW: Something I'd like to see on vanilla's block and item classes is a pre(or post)fix for generic classes, those that are used for multiple different blocks/items instead for just one. e.g. BlockGenericColored (GenericColoredBlock). That way those are easy to find (and reuse).

**Comment by InsomniaKitten** [comment]
> We should kill TileEntity and go with what Mojang has named them since a long time - BlockEntity.

If we were to go and rename `TileEntity` to `BlockEntity`, I personally feel it would make a lot more sense as merely `BlockData` (or something similar), as it is literally just NBT storage. ITickable is what handles updates akin to an Entity. Without that implementation, it is merely an object for storing and syncing NBT. This is just a suggestion, feel free to ignore it :P

**Comment by LexManos** [comment]
>Things that have been discussed and rejected:
TileEntity->BlockEntity. I know everyone's arguments for and against this. It is of my opinion that TileEntity is such a deeply ingrained name in the Minecraft modding community that changing it would not be worth it.

**Comment by williewillus** [comment]
@InsomniaKitten `BlockEntity` is pretty much confirmed to be the real Notch name so it's really between `BlockEntity` and `TileEntity`, nothing else

**Comment by KingLemming** [comment]
I'm fine with BlockEntity. I would definitely like it if the TE abbreviation only meant my mod. :P

**Comment by Pokechu22** [comment]
If `TileEntity` is being moved to a suffix, I'd say change it to `BlockEntity` in the same process; otherwise IMO leave it alone.

**Comment by DarkGuardsman** [comment]
I'm game for BlockEntity as well, in the long-term I think this will reduce confusion. Since it would be obvious that it's directly tied to the block. 

As for Prefix vs Sufix... for the love of all sanity do not change this just to improve searching. I really do not want to port over 200+ repositories over a change that has very little impact. I'm already heavily behind on updates and making this change will cause a lot of modders grief.

**Comment by C6H2Cl2** [comment]
I think BlockEntity is so bad name for modding. It is little hard to distinguish BlockEntity class from its Block class. I use ‘Tile’ prefix for all my TileEntity class because TileEntity is too long for prefix, but I can’t do same thing for BlockEntity. TileEntity is called TE for short, and BlockEntity will be called BE in the same way. BE is same for ‘BedrockEdition’, it is so bad.

**Comment by DarkGuardsman** [comment]
That is a good point, I keep forgetting a lot of people use the prefix Tile in place of TileEntity. Not sure how I missed that considering I use to do the same thing before switching over to my Node system.

**Comment by liach** [comment]
`EnchantmentType` should not be `TypeEnchantment`. In general, I agree with suffixing. For example, `WorldServer` is a world instead of a server.

**Comment by liach** [comment]
And @kashike, you should have told me this change in #456

**Comment by darkhax** [comment]
I am not a fan of moving from prefix to suffix. Prefix's help in several sorting/lookup situations. This change feels like it would offer very little for the majority of mod devs and cause more confusion and ultimately more rejection towards future updates. I am in favour of removing the enum prefix completely and not adding it as a suffix. Lastly, with the TileEntity/BlockEntity debate. I think BlockEntity would be the better choice and I would be in favour of changing it, however given how prominent TileEntity is, I don't mind if it stays the same.

**Comment by williewillus** [comment]
yeah, the list needs some manual going over, there's quite a few derpy names due to regex

**Comment by PaleoCrafter** [comment]
I see a lot of comments in favour of prefix naming because it makes lookups easier. However, the very same argument can be made for suffix naming. It's merely a matter of which category (i.e. type like 'block' vs subject like 'redstone') you prefer for lookup.
I may want to find everything related to the term 'stone' and will now have a neatly aligned list (due to the 'Stone' prefix) which let's me easily focus on the actually relevant part of the name.

Basically, it boils down to an issue of IDE support. I know for a fact that IDEA supports 'camel-case' search, such that a lookup for 'SBlock' should serve the same purpose as 'BlockS' does now. That basically makes this a non-issue.

With regards to Henry's packaging and naming scheme: I personally don't see the problem with using suffixes (according to type) here. If all names start with 'Alloy' for instance, it IMHO is at least as easy to parse the names for relevant information (in this case the type) quickly, as you can easily ignore what is now the prefix, due to the above-mentioned alignment of names.

I'm well aware that the same arguments mostly apply for prefix naming, making the final decision come down to preference. I myself am in favour of the suffixes and as was already outlined by others, this decision is for Vanilla names only, so we should also consider it exclusively in the Vanilla context.

Besides our own Remapper, there's also IDEA's migration ability for fixing mods, but I'm not sure if there's an eclipse equivalent.

**Comment by Drullkus** [comment]
`BlockEntity` is definitely a more fitting name. People are not wrong though for wanting `TileEntity` to stay because it's the ingrained terminology. I would do agree with `BlockEntity`

**Comment by robotgryphon** [comment]
I agree with BlockEntity as it's official.

As far as game libraries go, I see suffixes used more and more- so things like WatcherBlock and FurnaceEntity being commonplace.

**Comment by DarkGuardsman** [comment]
I say we remove the argument of sorting from the list. Since it really does depend on what tools you're using and how you develop.

**Comment by raoulvdberge** [comment]
Let's not remove that argument. Default sorting works like that and not everyone feels like configuring that change in their IDE or explorer. Besides, many people are used to this way of naming now so I suggest we keep the status quo and not change prefix/suffixing.

**Comment by oMilkyy** [comment]
If 1.13 is going to be such a big change for modders why not set a new standard all together? The point about the prefix organising code is a good one, but I see the putting a collection of classes in their correct packages as a better solution as most people already do this. 

The TileEntity -> Tile argument doesn't really stand just suffix your classes with Tile?

Changing TileEntity -> BlockEntity makes more sense as there is no mention of "Tiles" anywhere else in Minecraft it's based around "Blocks" and anything that isn't a block if sufficiently named.

Intellij and Eclipse both support asterisks searches "*Block" will give you everything with "Block" in the name, and also as mentioned before camel-case searches "LB" will give you LogBlock

For the sorting argument in IDE having the class "type" as the suffix allows you to see all the related classes for the "Block", or "Entity" etc, you are looking for which isn't a bad thing.

As for the the enum and special case classes they should probably be left alone to save confusion like the EnchantmentType example. Just my take on this, sorry for annoying anyone with these.

**Comment by DarkGuardsman** [comment]
The reason I say remove the sorting argument is it not a valid reason to make a major change. Since it honestly is down to how each person develops and uses there IDE tools. Basing a major change on "it would make sorting or searching easier" is the same as saying "I would have an easier time". Since you can not say "everyone would have an easier time" or even "the majority would have an easier time". It really should be removed as its an opinion based change.

As for making changes because 1.13 would be a big change. We have went through a lot of large changes as modders. Each time its a pain to port over mods to match the changes. Due to this each major version mods are left behind as the changes are too much to deal with.

Take myself who maintains 200 repositories related to Minecraft. After the major render update in 1.8 I have been unable to keep up with updates. Making a major change on top of existing major changes will put someone like me every farther behind. I know I'm in edge case due to this but I've interacted with a lot of modders over the last 7 years. What I know is that modders do not like to spend time updating. So reducing the time it takes to update is very important.

Now moving things to packages is fine as imports are not to hard to fix. However, moving from a prefix so a sufix would require a refactor of almost every class in a mod that uses blocks & items. Then if someone wants to match the new standard a refactor of every class name. This is on top of making changes for anything that would have normally changed between 1.12 to 1.13.

**Comment by raoulvdberge** [comment]
Agreed. The change is pointless because it comes down to preference. So, keeping status quo is the best you can do.

**Comment by maruohon** [comment]
In general the occasional class, method or field renames are welcome, as they **improve** the name - the new name may better describe what the class is or does, the old name might have been completely wrong, or the old name might have been unnecessarily long with filler words.

This Block and Item prefix -> suffix change though doesn't seem to objectively *improve* anything. It's only a naming convention change. There are lots of people who prefer one or the other. One or the other may be better in different situations, depending on what and how you are viewing/sorting.

We are writing code here, not a novel. The names don't need to read as perfect English. Grouping and categorizing though are important (at least for me) for code readability and the ability to see and find things quickly. Prefixed names keep things grouped and organized. Suffixing doesn't group anything, and the suffix is therefore almost useless in this regard.

In my opinion, this change would cause perhaps considerable amounts of extra work for everyone, just to have a different naming convention, that doesn't actually improve the names themselves in any way. This is basically forcing some people's preference to all the others. And the key thing here is that making this change now, after like 8 years or whatever, seems like an extremely unnecessary change. It will just cause confusion when talking about things (as you would now have to always also mention the version you are talking about), break the naming convention that has been in use for countless years, and start annoying the other half of modders than what the existing naming convention does.

Basically don't fix what isn't broken, especially when the fix "breaks everything", and doesn't fix anything.

**Comment by liach** [comment]
Another example of prefix vs suffix: who likes `ListArray` over `ArrayList`? Leave a 👎  if you like `ListArray` more.

**Comment by DarkGuardsman** [comment]
> Basically don't fix what isn't broken, especially when the fix "breaks everything", and doesn't fix anything.
 👍 6  
That sums this entire argument up well as "If its not broken, don't fix it"

**Comment by maxanier** [comment]
I also don't see a real reason for the Prefix/Suffix renaming.
It would mostly cause confusion and force modders to readjust. If it would bring any benefit, ok, but I don't see any.

All the suggestions above how to use IDE tools to still be able to filter for blocks (CamelCase search, asterix, ...) when moving to suffixes also work the other way around (if you currently don't want to type the prefix).
If I currently start typing `Stone` in IntelliJ  it suggests all stone related classes. If I type `BStone` it suggests `BlockStone`. This should work regardless of prefixing or postfixing, so there is no real improvement of convenience. 

Regarding readability it is very similar I think.
When reading vanilla (or modded) code you sometimes wonder what kind of class something is first  so  `BlockStone` would be quicker and sometimes you already know it is a block so `StoneBlock` would improve readiblity.

So why not leave it as it is.

**Comment by DarkGuardsman** [comment]
I thought about pointing out readability earlier but avoided it as it is down to coding methodology. Personally, I love to do all my classes as TypeName (BlockStone in your example). As you noted this allows a developer to quickly understand what type the class is before even looking at the name. In my opinion this is very import as the name can be any length. 

Image in the case of TheNameOfMySpecialCustomBlock. It reads well in english but I can't tell what it is without opening the class or seeing block at the end.

**Comment by sp614x** [comment]
I am strongly against the mass renaming `prefix -> suffix`. For me it has no real advantages and a lot of  disadvantages.

1. It would make updating from 1.12 to 1.13 much more complex and error prone.
2. It would make porting future development to older versions practically impossible.
3. It would make following the code history much harder.

Many mods and players are staying on older editions (1.7.10, 1.8, 1.8.9, ...) for various reasons. Because of this porting new development and bugfixes to older versions is becoming more and more important. The internal changes in 1.13 would make the backporting harder, similar to 1.8. Adding the mass renaming on top of it would make backporting practically impossible. 

A stable "non-optimal" base is much better than a "perfect" permanently shifting base.

**Comment by DarkGuardsman** [comment]
To add to sp614x's comment: 
I can say from experience that permanently shifting base destroys progress. While working for the Universal Electricity team we changed the base and API too often. Developers that tried to support our API left. Even though we had one of the best power APIs when comparing for performance and flexibility. Too many changes make it unrealistic to maintain versions. It got so bad our team fell apart due to the lack of focus and constant changes.

Tbh this feels like what keeps happening to forge. We do more than necessary and less than is needed each update.  I think the primary focus should always be on decreasing workload for mod developers. This way developers stick around, new developers learn faster, and developers can focus on content rather than updates. The more we change the less this happens. This can be seen just by how many more mods exist for 1.7.10 vs any other newer version.

**Comment by Boundarybreaker** [comment]
I feel having `Block`, `Item`, `Material`, and the like makes a lot more sense because of the sort of organization Java tends to have. It really tends to prefixes more than suffixes overall, what with the Java identifier scheme being `tld.domain.project` and the like. If they were used as a suffix, it would kinda feel like the name of the thing is the prefix instead, and it just wouldn't fit in. Especially with TileEntities, prefixing with `TileEntity` in the case of a `Render` but suffixing it for the TileEntity itself just doesn't make sense. (Side note, I *really* like the `TileEntity<Name>Render` for renders, since they're pretty much independent of the TileEntity. That's what a suffix really feels like it is; something that's dependent on what comes before it.)

**Comment by oMilkyy** [comment]
Most java classes are suffixed with their "type", ArrayList, HashMap, HashSet are all suffixed and so are alot of other classes just using them for examples

**Comment by tyra314** [comment]
I'm against renaming at all. Some aspects have already been discussed, but I'd like to add one more thing: **Stability**.

Breaking the interface now means all existing mods must be adopted, and there are *a lot* of mods. Breaking the interface is fine, as long as there are good reasons such as changes to vanilla's core. But breaking the interface because something is called wrong, is simply not worth it. I can totally understand the thrive for perfecting the code base and having THE perfect name for everything. But that is time spent on renaming things in too much existing code. Sure, there are tools for that, but it still has to be done by someone. I'd rather spend my time on other things than renaming stuff or learning how to bulk rename stuff. And there is another related aspect: 

Getting into Minecraft modding is already hard enough. The documentation is spread all over the internet in different states of obsolescence. But there is tons of good material out there. Renaming invalidates all these resources. You might be able to fix from source generated documentation with tools, possibly even wikis, but good look fixing videos or podcasts. 

Overall, I really appreciate your hard work, but I don't think the cause is worth the trouble it makes.
